VideoReceiverApp/main.cpp

```diff
@@ -299,52 +302,27 @@ VideoReceiverApp::exec()
 
     startStreaming();
 
-    QObject::connect(_receiver, &VideoReceiver::timeout, [this](){
+    QObject::connect(_receiver, &VideoReceiver::timeout, [](){
         qCDebug(AppLog) << "Streaming timeout";
-
-        _dispatch([this](){
-            if (_receiver->streaming()) {
-                _receiver->stop();
-            } else {
-                if (--_connect > 0) {
-                    qCDebug(AppLog) << "Restarting streaming";
-                    _dispatch([this](){
-                        startStreaming();
-                    });
-                } else {
-                    qCDebug(AppLog) << "Closing...";
-                    delete _receiver;
-                    _app.exit();
-                }
-            }
-        });
      });
 
-    QObject::connect(_receiver, &VideoReceiver::streamingChanged, [this](){
-        if (_receiver->streaming()) {
+    QObject::connect(_receiver, &VideoReceiver::streamingChanged, [this](bool active){
+        _streaming = active;
+        if (_streaming) {
             qCDebug(AppLog) << "Streaming started";
         } else {
             qCDebug(AppLog) << "Streaming stopped";
-            _dispatch([this](){
-                if (--_connect > 0) {
-                    qCDebug(AppLog) << "Restarting streaming";
-                    startStreaming();
-                } else {
-                    qCDebug(AppLog) << "Closing...";
-                    delete _receiver;
-                    _app.exit();
-                }
-            });
         }
      });
 
-    QObject::connect(_receiver, &VideoReceiver::decodingChanged, [this](){
-        if (_receiver->decoding()) {
+    QObject::connect(_receiver, &VideoReceiver::decodingChanged, [this](bool active){
+        _decoding = active;
+        if (_decoding) {
```

Comment:
1 - Change the outer lambda to a method, call it `toggleDecoding(bool toggle)`

The method is too big to be in a lambda, specially one that has lambdas inside.

2 - Always pass `this` as the third element of the connect, this makes sure that when a object is deleted the connection is broken and crashes will not happen because we are not tryng to trigger a callback on a deleted object.

`connect(_receiver, &VideoReceiver::decodingChanged, this, [this](...){});`

(see the two `this` there)

3 - Implement with the `failure first / bail out early` approach - your code always have too many indentation levels and it's hard to keep in mind what's happening, specially in code that will call itself in another thread.
```

_decoding = active;
if (_decoding) {
   qCDebug(AppLog) << "Decoding started";
   return; <- no need for the else, less chain if's.
}
```

4 -  store the lambda in a variable before  calling lambdas within lambdas. When we are reading code fast sometimes  we ignore certain symbols and we will misunderstand code [] can be lambda and array access, easy to miss when you are tired.
```

qCDebug(AppLog) << "Decoding Stopped");
if (_streaming && !_recording) {    
    auto stopLambda = [this] { _receiver->stop(); }; 
    _dispatch(lambda);
}
```